<a href="https://colab.research.google.com/github/Malinowsk/Optativa-IA/blob/main/Rago_Juan_Ignacio_Eugenio_Romanelli_TP_Reglas_de_asociaci%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **TP Reglas de Asociación**

Cargamos la librería que nos permite acceder al dataset

In [ ]:
# soporte para cargar dataset de https://www.openml.org/
!pip install openml
import openml

Accedemos al dataset

In [ ]:
import pandas as pd

# indicamos cual dataset queremos utilizar, en este caso el nro. 24
dataset = openml.datasets.get_dataset(24)

# separamos las información almacenada en el dataset
X, y, categorical_indicator, attribute_names = dataset.get_data(
    dataset_format='dataframe',
    target=dataset.default_target_attribute
)

#  concatenamos la información relevante en un único DataFrame 
df = pd.concat([X, y], axis=1)
df

cap-shape cap-surface cap-color bruises%3F odor gill-attachment  \
0            x           s         n          t    p               f   
1            x           s         y          t    a               f   
2            b           s         w          t    l               f   
3            x           y         w          t    p               f   
4            x           s         g          f    n               f   
...        ...         ...       ...        ...  ...             ...   
8119         k           s         n          f    n               a   
8120         x           s         n          f    n               a   
8121         f           s         n          f    n               a   
8122         k           y         n          f    y               f   
8123         x           s         n          f    n               a   

     gill-spacing gill-size gill-color stalk-shape  ...  \
0               c         n          k           e  ...   
1               c         b          k           e  ...   
2               c         b          n           e  ...   
3               c         n          n           e  ...   
4               w         b          k           t  ...   
...           ...       ...        ...         ...  ...   
8119            c         b          y           e  ...   
8120            c         b          y           e  ...   
8121            c         b          n           e  ...   
8122            c         n          b           t  ...   
8123            c         b          y           e  ...   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat class  
0              o         p                 k          s       u     p  
1              o         p                 n          n       g     e  
2              o         p                 n          n       m     e  
3              o         p                 k          s       u     p  
4              o         e                 n          a       g     e  
...          ...       ...               ...        ...     ...   ...  
8119           o         p                 b          c       l     e  
8120           o         p                 b          v       l     e  
8121           o         p                 b          c       l     e  
8122           o         e                 w          v       l     p  
8123           o         p                 o          c       l     e  

[8124 rows x 23 columns]

**a) Describir el dataset seleccionado.**
>El dataset seleccionado contiene información sobre las caracteristicas fisicas de los hongos, clasificandolos en: venenosos o comestibles. Con un total de 23 caracteristicas para 8124 ejemplares de hongos.

>Este dataset corresponde a registros de hongos extraidos de The Audubon Society Field Guide to North American Mushrooms (1981).

**b) Definir cuál es el objetivo de aplicar la técnica de Reglas de Asociación en el dataset elegido. Explicar qué tipo de patrones se esperan encontrar.** 
> El objetivo es descubrir patrones de atributos que suelen ocurrir juntos. Se espera encontrar patrones en los cuales los hongos que son venenesos posean determinadas caracteristicas y los no venenosos posean otras.

**d) Definir qué pre‐procesamiento se realizará a los datos. Justificar. Aplicarlo.**
> El dataset de entrada que debemos proveerle a la función *apriori* debe tener un formato particular llamado "One-hot encoded", para ello debemos realizar un pre-procesamiento utilizando la funcion get_dummies() provista por Pandas.

> Utilizaremos un soprte con valor de 0.3 ya que tenemos muchas transacciones y muchos items, utilizando valores desoporte altos no obtendriamos muchos itemsets frecuentes.



In [ ]:
from mlxtend.frequent_patterns import apriori

#Formato One-hot
df_dummies = pd.get_dummies(df , columns = df.columns)
#print(df_dummies)

#Itemsets frecuentes y soportes asociado.
frequent_itemsets = apriori(df_dummies, min_support=0.3, use_colnames=True )

#Agragamos una columna que almacena la longitud de cada itemset.
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))

In [ ]:
#Seleccionamos los itemsets frecuentes que contengan las clases de hongos venenosos.
frequent_itemsets[ frequent_itemsets['itemsets'].apply(lambda x: 'class_p' in x)]

support                                           itemsets  length
26    0.482029                                          (class_p)       1
53    0.405219                            (bruises%3F_f, class_p)       2
91    0.479813                       (class_p, gill-attachment_f)       2
107   0.468242                          (gill-spacing_c, class_p)       2
173   0.482029                             (class_p, veil-type_p)       2
...        ...                                                ...     ...
2526  0.302314  (gill-spacing_c, ring-number_o, gill-attachmen...       7
2528  0.308223  (ring-number_o, gill-attachment_f, veil-color_...       7
2529  0.302314  (gill-spacing_c, ring-number_o, veil-color_w, ...       7
2668  0.335795  (gill-spacing_c, ring-number_o, gill-attachmen...       7
2691  0.302314  (gill-spacing_c, ring-number_o, gill-attachmen...       8

[128 rows x 3 columns]

**e) Obtener las reglas de asociación utilizando el dataset elegido. Definir parámetros del algoritmo. Justificar.**
>Para generar las reglas de asociacion utilizaremos la funcion *associations_rules* pasando como parámetro los itemsetes frecuentes generados, la metrica a utilizar (en nuestro caso será "confidence"), y el valor minimo de confianza (1).


In [ ]:
from mlxtend.frequent_patterns import association_rules

# utilizamos confianza = 1
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.9)
rules

antecedents  \
0                                          (cap-shape_f)   
1                                          (cap-shape_f)   
2                                          (cap-shape_f)   
3                                          (cap-shape_f)   
4                                          (cap-shape_x)   
...                                                  ...   
24402  (bruises%3F_t, stalk-surface-below-ring_s, vei...   
24403  (bruises%3F_t, stalk-surface-above-ring_s, sta...   
24404  (bruises%3F_t, stalk-surface-below-ring_s, vei...   
24405  (bruises%3F_t, ring-number_o, stalk-surface-be...   
24406  (bruises%3F_t, stalk-surface-below-ring_s, rin...   

                                             consequents  antecedent support  \
0                                    (gill-attachment_f)            0.387986   
1                                          (veil-type_p)            0.387986   
2                                         (veil-color_w)            0.387986   
3                                        (ring-number_o)            0.387986   
4                                    (gill-attachment_f)            0.450025   
...                                                  ...                 ...   
24402  (gill-spacing_c, ring-number_o, gill-attachmen...            0.350566   
24403  (gill-spacing_c, ring-number_o, gill-attachmen...            0.341704   
24404  (gill-spacing_c, ring-number_o, gill-attachmen...            0.350566   
24405  (gill-spacing_c, gill-attachment_f, veil-color...            0.337765   
24406  (gill-spacing_c, ring-number_o, gill-attachmen...            0.350566   

       consequent support   support  confidence      lift  leverage  \
0                0.974151  0.381339    0.982868  1.008949  0.003382   
1                1.000000  0.387986    1.000000  1.000000  0.000000   
2                0.975382  0.381832    0.984137  1.008976  0.003397   
3                0.921713  0.371738    0.958122  1.039501  0.014126   
4                0.974151  0.443378    0.985230  1.011373  0.004986   
...                   ...       ...         ...       ...       ...   
24402            0.484490  0.316100    0.901685  1.861101  0.146254   
24403            0.772033  0.316100    0.925072  1.198228  0.052294   
24404            0.484490  0.316100    0.901685  1.861101  0.146254   
24405            0.385032  0.316100    0.935860  2.430603  0.186050   
24406            0.484490  0.316100    0.901685  1.861101  0.146254   

       conviction  
0        1.508835  
1             inf  
2        1.551945  
3        1.869388  
4        1.750096  
...           ...  
24402    5.243469  
24403    3.042476  
24404    5.243469  
24405    9.587910  
24406    5.243469  

[24407 rows x 9 columns]

**f) Determinar qué post‐procesamiento se realizará sobre las reglas obtenidas. Aplicarlo.**
>Para filtrar las reglas que son de nuestro interés utilizaremos las herramientas que prevee la libreria Pandas.
>En nuestro caso buscamos reglas que en su consecuente tengan class_p, lo cual indica que es un hongo venenoso.


Agregamos una columna con la longitud del antecedente

In [ ]:
rules["antecedent_len"] = rules["antecedents"].apply(lambda x: len(x))
rules

In [ ]:
rules[(rules['consequents'] == {'class_p'})]

>Seleccionamos las reglas que poseen al menos tres items en su antecedente, y un lift minimo de 1.2.

In [ ]:
rules[ (rules['antecedent_len'] >= 2) &
      (rules['lift'] > 1.8) &
      (rules['consequents'] == {'class_p'})]

antecedents consequents  \
206                       (gill-spacing_c, bruises%3F_f)   (class_p)   
238                         (bruises%3F_f, population_v)   (class_p)   
1083   (gill-spacing_c, bruises%3F_f, gill-attachment_f)   (class_p)   
1160     (bruises%3F_f, population_v, gill-attachment_f)   (class_p)   
1172         (gill-spacing_c, bruises%3F_f, veil-type_p)   (class_p)   
1183        (gill-spacing_c, bruises%3F_f, veil-color_w)   (class_p)   
1190       (gill-spacing_c, bruises%3F_f, ring-number_o)   (class_p)   
1194        (gill-spacing_c, bruises%3F_f, population_v)   (class_p)   
1261           (bruises%3F_f, population_v, veil-type_p)   (class_p)   
1277          (bruises%3F_f, population_v, veil-color_w)   (class_p)   
1281         (bruises%3F_f, population_v, ring-number_o)   (class_p)   
3714   (gill-spacing_c, bruises%3F_f, veil-type_p, gi...   (class_p)   
3736   (gill-spacing_c, bruises%3F_f, gill-attachment...   (class_p)   
3753   (gill-spacing_c, ring-number_o, gill-attachmen...   (class_p)   
3763   (gill-spacing_c, bruises%3F_f, population_v, g...   (class_p)   
3912   (veil-type_p, population_v, gill-attachment_f,...   (class_p)   
3947   (bruises%3F_f, population_v, gill-attachment_f...   (class_p)   
3955   (ring-number_o, population_v, gill-attachment_...   (class_p)   
3978   (gill-spacing_c, bruises%3F_f, veil-type_p, ve...   (class_p)   
3996   (gill-spacing_c, ring-number_o, veil-type_p, b...   (class_p)   
4005   (gill-spacing_c, bruises%3F_f, population_v, v...   (class_p)   
4018   (gill-spacing_c, ring-number_o, veil-color_w, ...   (class_p)   
4028   (gill-spacing_c, bruises%3F_f, population_v, v...   (class_p)   
4035   (gill-spacing_c, ring-number_o, population_v, ...   (class_p)   
4132   (bruises%3F_f, population_v, veil-type_p, veil...   (class_p)   
4140   (ring-number_o, population_v, veil-type_p, bru...   (class_p)   
4148   (ring-number_o, population_v, veil-color_w, br...   (class_p)   
8788   (gill-spacing_c, gill-attachment_f, veil-color...   (class_p)   
8824   (gill-spacing_c, ring-number_o, gill-attachmen...   (class_p)   
8845   (gill-spacing_c, gill-attachment_f, population...   (class_p)   
8874   (gill-spacing_c, ring-number_o, gill-attachmen...   (class_p)   
8896   (gill-spacing_c, gill-attachment_f, veil-color...   (class_p)   
8911   (gill-spacing_c, ring-number_o, gill-attachmen...   (class_p)   
9112   (gill-attachment_f, veil-color_w, population_v...   (class_p)   
9127   (ring-number_o, gill-attachment_f, population_...   (class_p)   
9145   (ring-number_o, gill-attachment_f, veil-color_...   (class_p)   
9180   (gill-spacing_c, ring-number_o, veil-color_w, ...   (class_p)   
9201   (gill-spacing_c, veil-color_w, population_v, b...   (class_p)   
9216   (gill-spacing_c, ring-number_o, population_v, ...   (class_p)   
9232   (gill-spacing_c, ring-number_o, veil-color_w, ...   (class_p)   
9287   (ring-number_o, veil-color_w, population_v, br...   (class_p)   
15357  (gill-spacing_c, ring-number_o, gill-attachmen...   (class_p)   
15402  (gill-spacing_c, gill-attachment_f, veil-color...   (class_p)   
15433  (gill-spacing_c, ring-number_o, gill-attachmen...   (class_p)   
15467  (gill-spacing_c, ring-number_o, gill-attachmen...   (class_p)   
15578  (ring-number_o, gill-attachment_f, veil-color_...   (class_p)   
15617  (gill-spacing_c, ring-number_o, veil-color_w, ...   (class_p)   
20908  (gill-spacing_c, ring-number_o, gill-attachmen...   (class_p)   

       antecedent support  consequent support   support  confidence      lift  \
206              0.435746            0.482029  0.392418    0.900565  1.868281   
238              0.334810            0.482029  0.308223    0.920588  1.909821   
1083             0.409897            0.482029  0.390202    0.951952  1.974887   
1160             0.322994            0.482029  0.308223    0.954268  1.979692   
1172             0.435746            0.482029  0.392418    0.900565  1.868281   
1183             0.412112     

**g) Analizar las reglas de asociación obtenidas.**

bbb

**h) Mostrar un ejemplo de cómo los resultados obtenidos pueden ser usados en el futuro.**

lkkkgfg